lib/net/scp.rb

```diff
@@ -362,7 +362,18 @@ def start_command(mode, local, remote, options={}, &callback)
               channel[:stack   ] = []
               channel[:error_string] = ''
 
-              channel.on_close                  { |ch2| send("#{channel[:state]}_state", channel); raise Net::SCP::Error, "SCP did not finish successfully (#{channel[:exit]}): #{channel[:error_string]}" if channel[:exit] != 0 }
+              channel.on_close do
+                if channel[:exit].nil? && channel[:state] == :finish
+                  # The remote closed the channel without sending an eof, but
```

Comment:
I'm a bit confused, but is it eof or exit-status?

https://www.ietf.org/rfc/rfc4254.txt
> 6.10.  Returning Exit Status
> 
>    When the command running at the other end terminates, the following
>    message can be sent to return the exit status of the command.
>    Returning the status is RECOMMENDED.  No acknowledgement is sent for
>    this message.  The channel needs to be closed with
>    SSH_MSG_CHANNEL_CLOSE after this message.
> 
>    The client MAY ignore these messages.
> 
>       byte      SSH_MSG_CHANNEL_REQUEST
>       uint32    recipient channel
>       string    "exit-status"
>       boolean   FALSE
>       uint32    exit_status
> 
>    The remote command may also terminate violently due to a signal.
>    Such a condition can be indicated by the following message.  A zero
>    'exit_status' usually means that the command terminated successfully.
> 
>       byte      SSH_MSG_CHANNEL_REQUEST
>       uint32    recipient channel
>       string    "exit-signal"
>       boolean   FALSE
>       string    signal name (without the "SIG" prefix)
>       boolean   core dumped
>       string    error message in ISO-10646 UTF-8 encoding
>       string    language tag [RFC3066]
>

Reply:
You are absolutely right. Not receiving SSH_MSG_CHANNEL_EOF is not the problem, I messed things up.

To clarify:
The remote implementation closes the channel without sending an eof nor an exit_status.
NET::SCP currently needs an exit_status = 0 to successfully end the transmission,  which causes my problem.
As RFC4254 specifies sending exit_status als RECOMMENDED, NET::SCP should cope with not getting one.

I will correct my code and submit again.